src/hooks/protections/handle_join_request.cpp

```diff
@@ -10,19 +11,41 @@ namespace big
 {
 	bool hooks::handle_join_request(Network* network, rage::snSession* session, rage::rlGamerInfo* player_info, CJoinRequestContext* ctx, BOOL is_transition_session)
 	{
-		if (auto player = g_player_database_service->get_player_by_rockstar_id(player_info->m_gamer_handle.m_rockstar_id);
-		    player && player->block_join) [[unlikely]]
-		{
+		auto rockstar_id = player_info->m_gamer_handle.m_rockstar_id;
+		auto is_friend   = friends_service::is_friend(rockstar_id);
+		auto db_player   = g_player_database_service->get_player_by_rockstar_id(rockstar_id);
+		auto block_join  = db_player && db_player->block_join;
+		auto is_trusted  = db_player && db_player->is_trusted;
+
+		auto send_response = [ctx](int block_join_reason = 1) {
 			CMsgJoinResponse response{};
-			response.m_status_code = player->block_join_reason;
+			response.m_status_code = block_join_reason;
 			g_pointers->m_gta.m_write_join_response_data(&response, ctx->m_join_response_data, 512, &ctx->m_join_response_size);
+		};
+
+		if (block_join) [[unlikely]]
+		{
+			send_response(db_player->block_join_reason);
 			g_notification_service.push("BLOCK_JOIN"_T.data(),
-			    std::vformat("BLOCK_JOIN_INFO"_T, std::make_format_args(player->name)));
+			    std::vformat("BLOCK_JOIN_INFO"_T, std::make_format_args(player_info->m_name)));
 			return false;
 		}
-		else [[likely]]
+		else if (g.session.lock_session) [[unlikely]]
 		{
-			return g_hooking->get_original<hooks::handle_join_request>()(network, session, player_info, ctx, is_transition_session);
+			if ((is_friend && g.session.allow_friends_into_locked_session) || is_trusted)
```

Comment:
Not required but [[unlikely]] and [[likely]] could probably be used. May want to wait for yimura or gir to approve that idea so it isn't a waste of time.

Reply:
This references a configuration boolean, and as such we cannot make assumptions about what the user will do, thus we can't make assumptions for the compiler as which path will be mostly taken. We made a core set of rules about when to use and when not to use the branch prediction adjustor modifiers in the original ticket we submitted the change in. https://github.com/YimMenu/YimMenu/pull/2929#issuecomment-2045245855